[loki] NullPointerException in GroupConsumerRetry

Reported automatically by loki-monitoring error triage.

- **Fingerprint:** `10374662eb26`
- **Total occurrences:** 1 (over 1 runs)
- **First seen:** 2026-07-03T10:10:38.074Z
- **Last seen:** 2026-07-03T10:10:38.074Z
- **Level:** ERROR
- **Logger:** `org.apache.james.events.GroupConsumerRetry`

### Message

Exception happens when handling event after 0 retries

### Stacktrace

```
java.lang.NullPointerException: Cannot invoke "com.google.firebase.messaging.MessagingErrorCode.name()" because the return value of "com.google.firebase.messaging.FirebaseMessagingException.getMessagingErrorCode()" is null
	at com.linagora.tmail.james.jmap.firebase.FirebasePushListener.$anonfun$sendNotification$4(FirebasePushListener.scala:114)
	at reactor.core.scala.publisher.SMono.$anonfun$onErrorResume$1(SMono.scala:836)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:134)
	at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onError(MonoIgnoreElements.java:84)
	at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onError(MonoIgnoreElements.java:84)
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onError(MonoIgnoreThen.java:280)
	at reactor.core.publisher.MonoCreate$DefaultMonoSink.error(MonoCreate.java:205)
	at com.linagora.tmail.james.jmap.firebase.FirebasePushClient.lambda$sendReactive$1(FirebasePushClient.java:193)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1004)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:767)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:516)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.afterRanInterruptiblyFailure(TrustedListenableFutureTask.java:138)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:89)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:80)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	Suppressed: com.google.firebase.messaging.FirebaseMessagingException: Unknown error while making a remote service call: Connection reset
		at com.google.firebase.messaging.FirebaseMessagingException.withMessagingErrorCode(FirebaseMessagingException.java:51)
		at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:293)
		at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:282)
		at com.google.firebase.internal.AbstractHttpErrorHandler.handleIOException(AbstractHttpErrorHandler.java:63)
		at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:110)
		at com.google.firebase.internal.ErrorHandlingHttpClient.sendAndParse(ErrorHandlingHttpClient.java:72)
		at com.google.firebase.messaging.FirebaseMessagingClientImpl.sendSingleRequest(FirebaseMessagingClientImpl.java:127)
		at com.google.firebase.messaging.FirebaseMessagingClientImpl.send(FirebaseMessagingClientImpl.java:113)
		at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:141)
		at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:138)
		at com.google.firebase.internal.CallableOperation.call(CallableOperation.java:36)
		at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:128)
		at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
		... 4 common frames omitted
	Caused by: java.net.SocketException: Connection reset
		at java.base/sun.nio.ch.NioSocketImpl.implRead(Unknown Source)
		at java.base/sun.nio.ch.NioSocketImpl.read(Unknown Source)
		at java.base/sun.nio.ch.NioSocketImpl$1.read(Unknown Source)
		at java.base/java.net.Socket$SocketInputStream.implRead(Unknown Source)
		at java.base/java.net.Socket$SocketInputStream.read(Unknown Source)
		at java.base/java.io.BufferedInputStream.fill(Unknown Source)
		at java.base/java.io.BufferedInputStream.read1(Unknown Source)
		at java.base/java.io.BufferedInputStream.read(Unknown Source)
		at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(Unknown Source)
		at java.base/sun.net.www.http.HttpClient.parseHTTP(Unknown Source)
		at java.base/sun.net.www.protocol.http.HttpURLConnection.doTunneling0(Unknown Source)
		at java.base/sun.net.www.protocol.http.HttpURLConnection.doTunneling(Unknown Source)
		at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
		at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
		at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
		at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(Unknown Source)
		at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:113)
		at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:84)
		at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1012)
		at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:96)
		... 12 common frames omitted
```

<sub>loki-errors-fingerprint: 10374662eb26</sub>

--- a/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/firebase/FirebasePushListener.scala
+++ b/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/firebase/FirebasePushListener.scala
@@ -111,7 +111,7 @@ class FirebasePushListener @Inject()(subscriptionRepository: FirebaseSubscriptio
             case _ =>
               pushFailure.increment()
               SMono.error(new RuntimeException(
-                s"Unexpected error during push message to Firebase Cloud Messaging for user ${stateChangeEvent.username.asString()} and subscription ${subscription.id.serialize} and code ${e.getMessagingErrorCode.name()}", e))
+                s"Unexpected error during push message to Firebase Cloud Messaging for user ${stateChangeEvent.username.asString()} and subscription ${subscription.id.serialize} and code ${Option(e.getMessagingErrorCode).map(_.name()).getOrElse("UNKNOWN")}", e))
           }
           case e =>
             pushFailure.increment()

--- a/tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/firebase/FirebasePushListenerTest.scala
+++ b/tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/firebase/FirebasePushListenerTest.scala
@@ -32,7 +32,7 @@ import org.apache.james.jmap.core.UuidState
 import org.apache.james.metrics.tests.RecordingMetricFactory
 import org.apache.james.user.api.DelegationStore
 import org.apache.james.user.memory.MemoryDelegationStore
-import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.{assertThat, assertThatThrownBy}
 import org.assertj.core.api.SoftAssertions
 import org.junit.jupiter.api.{BeforeEach, Test}
 import org.mockito.ArgumentCaptor
@@ -195,6 +195,30 @@ class FirebasePushListenerTest {
     SMono(testee.reactiveEvent(StateChangeEvent(EventId.random(), bob,
       Map(EmailTypeName -> UuidState(UUID.randomUUID())))))
       .onErrorResume(_ => SMono.empty).block()
+
+    assertThat(SMono.fromPublisher(subscriptionRepository.get(bob, java.util.Set.of(subscriptionId))).block())
+      .isNotNull
+  }
+
+  @Test
+  def shouldNotRemoveSubscriptionAndShouldNotThrowNpeWhenMessagingErrorCodeIsNull(): Unit = {
+    val firebaseException = mock(classOf[FirebaseMessagingException])
+    when(firebaseException.getMessagingErrorCode).thenReturn(null)
+    when(pushClient.push(any())).thenReturn(Mono.error(firebaseException))
+
+    val subscriptionId = SMono(subscriptionRepository.save(bob, FirebaseSubscriptionCreationRequest(
+      deviceClientId = DeviceClientId("junit"),
+      token = FirebaseToken("token"),
+      types = Seq(EmailTypeName)))).block().id
+
+    assertThat(SMono.fromPublisher(subscriptionRepository.get(bob, java.util.Set.of(subscriptionId))).block())
+      .isNotNull
+
+    assertThatThrownBy(() => SMono(testee.reactiveEvent(StateChangeEvent(EventId.random(), bob,
+      Map(EmailTypeName -> UuidState(UUID.randomUUID()))))).block())
+      .isInstanceOf(classOf[RuntimeException])
+      .hasMessageContaining("Unexpected error during push message to Firebase Cloud Messaging")
+      .hasCause(firebaseException)
 
     assertThat(SMono.fromPublisher(subscriptionRepository.get(bob, java.util.Set.of(subscriptionId))).block())
       .isNotNull

